src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs

```diff
@@ -894,29 +935,55 @@ internal virtual void PrepareSession()
                 WebSession.UserAgent = UserAgent;
             }
 
-            if (Proxy is not null)
+            // Proxy and NoProxy parameters are mutually exclusive.
+            // If NoProxy is provided, WebSession will turn off the proxy
+            // and if Proxy is provided NoProxy will be turned off.
+            if (NoProxy.IsPresent)
             {
-                WebProxy webProxy = new(Proxy);
-                webProxy.BypassProxyOnLocal = false;
-                if (ProxyCredential is not null)
-                {
-                    webProxy.Credentials = ProxyCredential.GetNetworkCredential();
-                }
-                else if (ProxyUseDefaultCredentials)
+                WebSession.NoProxy = true;
+            }
+            else
+            {
+                if (Proxy is not null)
                 {
-                    // If both ProxyCredential and ProxyUseDefaultCredentials are passed,
-                    // UseDefaultCredentials will overwrite the supplied credentials.
-                    webProxy.UseDefaultCredentials = true;
+                    WebProxy webProxy = new(Proxy);
+                    webProxy.BypassProxyOnLocal = false;
+                    if (ProxyCredential is not null)
+                    {
+                        webProxy.Credentials = ProxyCredential.GetNetworkCredential();
+                    }
+                    else if (ProxyUseDefaultCredentials)
+                    {
+                        // If both ProxyCredential and ProxyUseDefaultCredentials are passed,
+                        // UseDefaultCredentials will overwrite the supplied credentials.
+                        webProxy.UseDefaultCredentials = true;
+                    }
+
+                    // We don't want to update the WebSession unless the proxies are different
+                    // as that will require us to create a new HttpClientHandler and lose connection
+                    // persistence.
+                    if (!webProxy.Equals(WebSession.Proxy))
+                    {
+                        WebSession.Proxy = webProxy;
+                    }
```

Comment:
WebRequestSession does not do the check internally because it is using shared code to check for object equality when setting the cookies, proxy, certificates etc. These are publically settable properties on WebSession so can be changed directly in script. In general, these must be updated if the object references aren't the same as the user can modify the properties of the referenced objects after they are set on the WebSession. The WebProxy class is internal so can't easily be constructed by the user, and is a type used directly only by the WebRequestPSCmdlet so it makes sense to add the special case optimisation for not updating the internal proxy class if the cmdlet proxy parameters are identical from invocation to invocation in the class that directly knows about the equality optimisation.

The alternative if to use .Equals() for all reference types in the SetClassVar() function in WebRequestSession, which would be problematic if the equality operator says different references are Equal, but the user tries to modify an object after setting it in the WebRequestSession and it is ignored because of the Equality check.

Reply:
I agree to keep this as-is.
Not clear how underlying .Net API responds to changes contents of reference object parameters.